docs/proposals/cloud-edge-custom-message-design.md

```diff
@@ -0,0 +1,429 @@
+
+create-rule-rest-eventbus.yaml's contents are as follows:
+
+```
+apiVersion: rules.kubeedge.io/v1
+kind: Rule
+metadata:
+  name: my-rule
+  labels:
+    description: test
+spec:
+  source: "my-rest"
+  sourceResource: {"path":"/a"}
+  target: "my-eventbus"
+  targetResource: {"topic":"test"}
+status:
+  successMessages: 0
+  failMessages: 0
+  errors: []
+```
+
+1.3 call rest api in the cloud to send messages to edge.
+
+The rest api in the cloud can be called to send messages to eventbus on an edge node based on the node name and sourceResource. 
+
+- Method: POST
+- URL: **http://{rest_endpoint}/{node_name}/{namespace}/{path}**, {rest_endpoint} is router's endpoint, {node_name} is name of edgenode, {namespace} is the namespace of rule,{path} is source ruleEndpoint's sourceResource.
+- Body: {user_message}, {user_message} is user's message
+
+For example:
+- Method: POST
+- URL: http://{rest_endpoint}/{node_name}/default/a
+- Body： {"message":"123"}
+
+1.4 User's app subscribes custom topics from mqtt-broker in edge to receive messages from the cloud. 
+
+- Topic: {topic}, {topic} is target ruleEndpoint's targetResource.
+- Message:  {user_message}
+
+For example:
+- subscribe Topic: 'test', exec command with mosquitto: `mosquitto_sub -t 'test' -d`
+- Get Message: {"message":"123"}
+
+2. edge to cloud: **eventbus->rest**:
+
+2.1 create rest type and eventbus type ruleEndpoint if they don't exist. Exec command:
+
+	`kubectl create -f create-ruleEndpoint-rest.yaml;kubectl create -f create-ruleEndpoint-eventbus.yaml;`
+
+
+create-ruleEndpoint-rest.yaml's contents are as follows:
+```
+apiVersion: rules.kubeedge.io/v1
+kind: RuleEndpoint
+metadata:
+  name: my-rest
+  labels:
+    description: test
+spec:
+  ruleEndpointType: "rest"
+  properties: {}
+```
+
+create-ruleEndpoint-eventbus.yaml's contents are as follows:
+```
+apiVersion: rules.kubeedge.io/v1
+kind: RuleEndpoint
+metadata:
+  name: my-eventbus
+  labels:
+    description: test
+spec:
+  ruleEndpointType: "eventbus"
+  properties: {}
+```
+
+
+2.2 create rule. Exec command:
+
+`kubectl create -f create-rule-eventbus-rest.yaml`
+
+create-rule-eventbus-rest.yaml's contents are as follows:
+
+```
+apiVersion: rules.kubeedge.io/v1
+kind: Rule
+metadata:
+  name: my-rule-eventbus-rest
+  labels:
+    description: test
+spec:
+  source: "my-eventbus"
+  sourceResource: {"topic": "test","node_name": "xxx"}
+  target: "my-rest"
+  targetResource: {"resource":"http://a.com"}
+status:
+  successMessages: 0
+  failMessages: 0
+  errors: []
+```
+
+2.3 User's app in edge publishes messages with custom topic to MQTT broker on edge node. 
+- Topic: {namespace}/{topic}
+- Message:  {user_api_body}
+
+for example:
+- publish data with mosquitto, exec command:
+
+ `mosquitto_pub -t 'default/test' -d -m '{"edgemsg":"msgtocloud"}'`
+
+2.4 Kubeedge delivers messages to user api address in cloud. 
+
+- Method: POST
+- URL: **http://{user_api}**,{user_api} is target ruleEndpoint's targetResource.
```

Comment:
1. Should http be together with `{user_api}`?
2. Is only **http** supported?

Reply:
1.yes
2.support http and https, I will add https here.